spec/abci++/abci++_basic_concepts.md

```diff
@@ -0,0 +1,467 @@
+`Response*` or `Tx`. A `Response*` or `Tx` may contain multiple events with duplicate
+`type` values, where each distinct entry is meant to categorize attributes for a
+particular event. Every key and value in an event's attributes must be UTF-8
+encoded strings along with the event type itself.
+
+```protobuf
+message Event {
+  string                  type       = 1;
+  repeated EventAttribute attributes = 2;
+}
+```
+
+The attributes of an `Event` consist of a `key`, a `value`, and an `index` flag. The
+index flag notifies the Tendermint indexer to index the attribute. The value of
+the `index` flag is non-deterministic and may vary across different nodes in the network.
+
+```protobuf
+message EventAttribute {
+  bytes key   = 1;
+  bytes value = 2;
+  bool  index = 3;  // nondeterministic
+}
+```
+
+Example:
+
+```go
+ abci.ResponseFinalizeBlock{
+  // ...
+ Events: []abci.Event{
+  {
+   Type: "validator.provisions",
+   Attributes: []abci.EventAttribute{
+    abci.EventAttribute{Key: []byte("address"), Value: []byte("..."), Index: true},
+    abci.EventAttribute{Key: []byte("amount"), Value: []byte("..."), Index: true},
+    abci.EventAttribute{Key: []byte("balance"), Value: []byte("..."), Index: true},
+   },
+  },
+  {
+   Type: "validator.provisions",
+   Attributes: []abci.EventAttribute{
+    abci.EventAttribute{Key: []byte("address"), Value: []byte("..."), Index: true},
+    abci.EventAttribute{Key: []byte("amount"), Value: []byte("..."), Index: false},
+    abci.EventAttribute{Key: []byte("balance"), Value: []byte("..."), Index: false},
+   },
+  },
+  {
+   Type: "validator.slashed",
+   Attributes: []abci.EventAttribute{
+    abci.EventAttribute{Key: []byte("address"), Value: []byte("..."), Index: false},
+    abci.EventAttribute{Key: []byte("amount"), Value: []byte("..."), Index: true},
+    abci.EventAttribute{Key: []byte("reason"), Value: []byte("..."), Index: true},
+   },
+  },
+  // ...
+ },
+}
+```
+
+## Evidence
+
+[&uparrow; Back to Outline](#outline)
+
+Tendermint's security model relies on the use of "evidence". Evidence is proof of
+malicious behavior by a network participant. It is the responsibility of Tendermint
+to detect such malicious behavior. When malicious behavior is detected, Tendermint
+will gossip evidence of the behavior to other nodes and commit the evidence to
+the chain once it is verified by all validators. This evidence will then be
+passed on to the Application through ABCI++. It is the responsibility of the
+Application to handle the evidence and exercise punishment.
+
+EvidenceType has the following protobuf format:
+
+```protobuf
+enum EvidenceType {
+  UNKNOWN               = 0;
+  DUPLICATE_VOTE        = 1;
+  LIGHT_CLIENT_ATTACK   = 2;
+}
+```
+
+There are two forms of evidence: Duplicate Vote and Light Client Attack. More
+information can be found in either [data structures](../core/data_structures.md)
+or [accountability](../light-client/accountability/)
+
+## Errors
```

Comment:
* For `Code` they should be, as it is hashed... although the [documentation](https://github.com/tendermint/tendermint/blob/master/spec/core/data_structures.md#header) (in description of field `LastResultHash`), says that the `Codespace` is not hashed (this seems odd to me, but could be because `Codespace` was added later and they didn't want to break the block format to avoid a hard-fork)
* For `Query`, as it is only addressed to one node, I think it doesn't matter (no risk of diverging state machine replicas); although probably should conform to the App's spec (I mean the error has to make sense)